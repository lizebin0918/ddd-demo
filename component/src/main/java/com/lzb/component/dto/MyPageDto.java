package com.lzb.component.dto;

import java.util.List;
import java.util.function.Function;

import lombok.AllArgsConstructor;
import lombok.Getter;
import one.util.streamex.StreamEx;

/**
 * <br/>
 * Created on : 2023-09-22 23:07
 * @author lizebin
 */
@Getter
@AllArgsConstructor
public class MyPageDto<T> {

    private final long pages;
    private final long pageSize;
    private final long total;
    private final List<T> rows;

    private <E> MyPageDto(long pages, long pageSize, long total, List<E> rows, Function<E, T> convertor) {
        this.pages = pages;
        this.pageSize = pageSize;
        this.total = total;
        this.rows = StreamEx.of(rows).map(convertor).toList();
    }

    public static <E, T> MyPageDto<T> of(long pages,
            long pageSize,
            long total,
            List<E> rows,
            Function<E, T> convertor) {
        return new MyPageDto<>(pages, pageSize, total, rows, convertor);
    }
}
