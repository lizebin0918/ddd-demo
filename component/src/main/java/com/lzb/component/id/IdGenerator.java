package com.lzb.component.id;

/**
 * <br/>
 * Created on : 2023-09-08 13:10
 * @author lizebin
 */
public interface IdGenerator {

    String BEAN_NAME = "idGenerator";

    /**
     * 生成id
     * @return
     */
    long id();

}
