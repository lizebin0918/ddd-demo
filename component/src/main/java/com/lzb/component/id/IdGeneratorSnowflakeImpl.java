package com.lzb.component.id;

import cn.hutool.core.util.IdUtil;

import org.springframework.stereotype.Component;

/**
 * <br/>
 * Created on : 2023-09-08 13:11
 * @author lizebin
 */
@Component(IdGenerator.BEAN_NAME)
public class IdGeneratorSnowflakeImpl implements IdGenerator {

    public long id() {
        return IdUtil.getSnowflakeNextId();
    }
}
