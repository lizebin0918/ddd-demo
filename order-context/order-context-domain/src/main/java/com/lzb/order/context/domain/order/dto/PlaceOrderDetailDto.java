package com.lzb.order.context.domain.order.dto;

import java.math.BigDecimal;

import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;

/**
 * 生单指令<br/>
 * Created on : 2023-09-06 13:00
 * @author lizebin
 */
@Getter
@AllArgsConstructor
@EqualsAndHashCode
public class PlaceOrderDetailDto {

    @NotNull(message = "skuId cannot be empty")
    private final Integer skuId;

    @NotNull
    @Positive(message = "Quantity must be greater than 0")
    private final BigDecimal price;

    @NonNull
    @Positive(message = "Count must be greater than 0")
    private final Integer count;

}

/**
 * 另一种写法
 * @Getter
 * @EqualsAndHashCode
 * @SuperBuilder
 * @AllArgsConstructor
 * public class PlaceOrderDetailDto {
 *
 *     @NotBlank(message = "skuCode cannot be empty")
 *     private final String skuCode;
 *
 *     @NotNull
 *     @Positive(message = "Quantity must be greater than 0")
 *     private final Integer quantity;
 *
 * }
 */