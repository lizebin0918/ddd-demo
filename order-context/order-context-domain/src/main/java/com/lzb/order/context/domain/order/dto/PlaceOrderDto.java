package com.lzb.order.context.domain.order.dto;

import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotEmpty;
import java.math.BigDecimal;
import java.util.List;
import java.util.function.Predicate;

/**
 * 生单指令<br/>
 * Created on : 2023-09-06 13:00
 *
 * @author lizebin
 */
public record PlaceOrderDto(String currency, BigDecimal exchangeRate, BigDecimal totalShouldPay,
                            BigDecimal totalActualPay,
                            @Email(message = "邮箱格式不正确")
                            String email, String phoneNumber, String firstName,
                            String lastName, String addressLine1, String addressLine2, String country,
                            BigDecimal shipInsuranceFee,
                            Boolean isFirstOrder,
                            @NotEmpty
                            List<PlaceOrderDetailDto> details) {

    public boolean isOnlyOnePerSku(int maxSkuCount) {
        return details().stream().map(PlaceOrderDetailDto::getCount).allMatch(Predicate.isEqual(maxSkuCount));
    }
}
