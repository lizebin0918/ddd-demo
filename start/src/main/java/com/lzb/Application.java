package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.data.redis.RedisRepositoriesAutoConfiguration;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.scheduling.annotation.EnableAsync;

// 启动报错：Multiple Spring Data modules found, entering strict repository configuration
@SpringBootApplication(exclude = { RedisRepositoriesAutoConfiguration.class})
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
