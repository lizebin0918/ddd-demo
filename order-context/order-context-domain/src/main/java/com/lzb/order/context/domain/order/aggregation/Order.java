package com.lzb.order.context.domain.order.aggregation;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.function.IntUnaryOperator;
import java.util.stream.Collectors;

import cn.hutool.core.lang.Assert;
import com.lzb.component.aggregate.BaseAggregation;
import com.lzb.component.exception.BizException;
import com.lzb.order.context.domain.core.valobj.SkuCount;
import com.lzb.order.context.domain.inventory.entity.SkuStockLock;
import com.lzb.order.context.domain.order.aggregation.valobj.FullAddressLine;
import com.lzb.order.context.domain.order.aggregation.valobj.FullName;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderAddress;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderStatus;
import com.lzb.order.context.domain.order.dto.UpdateAddressDto;
import com.lzb.order.context.domain.order.event.OrderCanceledEvent;
import com.lzb.order.context.domain.order.event.OrderPlacedEvent;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
@Jacksonized
@SuperBuilder(toBuilder = true)
@Setter(AccessLevel.PACKAGE)
public class Order extends BaseAggregation<Order, Long> {

    private Long id;

    private OrderStatus orderStatus;

    private String currency;

    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    private OrderAddress orderAddress;

    private OrderDetails orderDetails;

    /**
     * 扩展字段
     */
    private Extension extension;

    @Builder.Default
    private Long userId = 1L;

    @NotNull
    public Set<Integer> getSkuIds() {
        return skuCounts().stream().map(SkuCount::skuId).collect(Collectors.toSet());
    }

    public void updateAddress(UpdateAddressDto updateOrderAddress) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        this.orderAddress = OrderAddress.builder()
            .fullName(FullName.of(updateOrderAddress.firstName(), updateOrderAddress.lastName()))
            .fullAddressLine(FullAddressLine.of(updateOrderAddress.addressLine1(), updateOrderAddress.addressLine2()))
            .country(updateOrderAddress.country())
            .email(updateOrderAddress.email())
            .phoneNumber(updateOrderAddress.phoneNumber())
            .build();
    }

    public void updateSkuLockStock(List<SkuStockLock> skuStockLocks) {
        IntUnaryOperator skuId2LockNum = lockedNum(skuStockLocks);
        orderDetails.forEach(orderDetail -> {
            orderDetail.updateLocked(orderDetail.getCount() <= skuId2LockNum.applyAsInt(orderDetail.getSkuId()));
        });
    }

    private IntUnaryOperator lockedNum(List<SkuStockLock> skuStockLocks) {
        Map<Integer, SkuStockLock> skuId2Locks = skuStockLocks.stream().collect(Collectors.toMap(SkuStockLock::skuId, Function.identity()));
        return skuId -> Optional.ofNullable(skuId2Locks.get(skuId)).map(SkuStockLock::lockedNum).orElse(0);
    }

    public void pay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = orderStatus.toCancel();
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return orderStatus.isCancel();
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        Assert.notBlank(fullName.getFirstName(), "firstName is null");
        this.orderAddress = orderAddress.toBuilder().fullName(fullName).build();
    }

    public boolean canCancel() {
        return false;
    }

    public void addOrderDetail(@NonNull OrderDetail orderDetail) {
        if (Objects.isNull(orderDetails)) {
            this.orderDetails = new OrderDetails(new ArrayList<>());
        }
        Set<Integer> inSkuIds = orderedSkuIds();
        if (inSkuIds.contains(orderDetail.getSkuId())) {
            throw new BizException("skuId重复");
        }
        orderDetails.add(orderDetail);
    }

    /**
     * 下单的skuIds
     * @return
     */
    private @NotNull Set<Integer> orderedSkuIds() {
        return orderDetails.toStream()
                .map(OrderDetail::getSkuId)
                .collect(Collectors.toSet());
    }

    public List<SkuCount> skuCounts() {
        return orderDetails.toStream().map(item -> new SkuCount(item.getSkuId(), item.getCount())).toList();
    }

    @Override
    public Long id() {
        return id;
    }
}
