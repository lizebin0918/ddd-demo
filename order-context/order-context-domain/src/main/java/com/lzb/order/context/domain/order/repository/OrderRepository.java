package com.lzb.order.context.domain.order.repository;

import java.util.List;
import java.util.Optional;

import com.lzb.component.repository.CacheRepository;
import com.lzb.component.repository.CommonRepository;
import com.lzb.order.context.domain.order.aggregation.Order;
import com.lzb.order.context.domain.order.aggregation.OrderReadonly;

/**
 * <br/>
 * Created on : 2023-08-30 23:07
 * @author lizebin
 */
public interface OrderRepository extends CommonRepository<Order, Long>, CacheRepository<Order, Long> {

    List<Order> list(long... orderIds);

    default Optional<OrderReadonly> getReadonly(long orderId) {
        return get(orderId).map(OrderReadonly::of);
    }

    default Optional<OrderReadonly> getReadonlyInCache(long orderId) {
        return getInCache(orderId).map(OrderReadonly::of);
    }
}
