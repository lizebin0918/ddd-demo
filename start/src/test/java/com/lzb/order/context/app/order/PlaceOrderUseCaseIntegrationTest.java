package com.lzb.order.context.app.order;

import java.math.BigDecimal;
import java.util.List;

import com.lzb.BaseIntegrationTest;
import com.lzb.component.id.IdGenerator;
import com.lzb.order.context.domain.order.dto.PlaceOrderDetailDto;
import com.lzb.order.context.domain.order.dto.PlaceOrderDto;
import com.lzb.order.context.domain.order.aggregation.Order;
import com.lzb.order.context.domain.order.repository.OrderRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;

/**
 * <br/>
 * Created on : 2023-09-09 14:23
 * @author lizebin
 */
class PlaceOrderUseCaseIntegrationTest extends BaseIntegrationTest {
    
    @Autowired
    private IdGenerator idGenerator;

    @Autowired
    private OrderAppService orderAppService;

    @Autowired
    private OrderRepository orderRepository;


    @Test
    @DisplayName("测试OrderDetailBuilder")
    void should_place_order() {

        doReturn(1L).when(idGenerator).id();

        PlaceOrderDto req = createPlaceOrderDto();

        long orderId = orderAppService.placeOrder(req);
        Order order = orderRepository.getOrThrow(orderId);
        assertJSON(order, "id");
    }

    private static PlaceOrderDto createPlaceOrderDto() {
        PlaceOrderDto req = new PlaceOrderDto("CNY", BigDecimal.ONE, BigDecimal.ONE, BigDecimal.ONE, "email",
                "phoneNumber", "firstName", "lastName", "addressLine1", "addressLine2", "country",
                BigDecimal.TEN, true, List.of(new PlaceOrderDetailDto(1, BigDecimal.ONE, 1)));
        return req;
    }

    @Test
    @DisplayName("测试缓存查询")
    void should_place_order_in_cache() {

        long orderId = 1L;
        doReturn(orderId).when(idGenerator).id();

        assertThat(orderRepository.getInCache(orderId).isEmpty()).isTrue();
        orderAppService.placeOrder(createPlaceOrderDto());
        assertThat(orderRepository.getInCache(orderId).isPresent()).isTrue();
    }


}