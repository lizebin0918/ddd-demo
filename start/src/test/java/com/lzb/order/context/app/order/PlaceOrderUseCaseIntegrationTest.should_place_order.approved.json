{
  "currency": "CNY",
  "exchangeRate": 1,
  "extension": {
    "isFirstOrder": true,
    "shipInsuranceFee": 10
  },
  "orderAddress": {
    "country": "country",
    "email": "email",
    "fullAddressLine": {
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2"
    },
    "fullName": {
      "firstName": "firstName",
      "lastName": "lastName"
    },
    "phoneNumber": "phoneNumber"
  },
  "orderDetails": {
    "list": [
      {
        "count": 1,
        "locked": true,
        "orderStatus": "WAIT_PAY",
        "price": 1,
        "skuId": 1,
        "version": 1
      }
    ]
  },
  "orderStatus": "WAIT_PAY",
  "totalActualPay": 1,
  "totalShouldPay": 1,
  "userId": 1,
  "version": 1
}